Add safe Memory::read and Memory::write methods

theduke opened the pull request:
Introduce `Memory::read` and `Memory::write` methods for accessing memory contents safely.

Also adds a test and mentions the  the new methods in the `Memory` docs.

Fixes #2484.

Currently most fallible methods use `anyhow::Error`. 
I have instead added a new `MemoryError`, since recovering from such errors is possible and it's more expressive.
(future proofed with `#[non_exhaustive]`).

Could to change it to a `anyhow::Error` if that's the preferred strategy (or wrap the MemoryError with anyhow).

I also already added a `MemoryError::Locked` variant, which is not required yet, but apparently will be once threading support lands. Happy to remove it or slap on a `#[doc(hidden)]` to prevent confusion.

alexcrichton commented on crates/wasmtime/src/externals.rs:
```diff
@@ -598,6 +598,29 @@ impl Table {
+    OutOfBounds,
+    /// The memory is already locked by another reader or writer.
+    // Note: currently unused.
+    Locked,
```
FWIW I don't think that we'll ever need this error because we'll basically be implementing `memory.copy` semantics which would never need an error for this.

theduke replied:
In that case ... should the error just be a struct? 

I don't see any other relevant error conditions.

alexcrichton replied:
That's true yeah, and I think we could go ahead and have it just be an opaque structure with a singular private field so we can add more to it later if we want.

theduke replied:
Done.

alexcrichton commented on crates/wasmtime/src/externals.rs:
```diff
@@ -864,6 +902,38 @@ impl Memory {
 
+    /// Copies memory contents at the given offset into a buffer.
+    ///
+    /// Returns an error if the copy failed.
```
Could this documentation detail the error cases as well? It also might be good to mention that the size of the copy is the size of the buffer itself.

theduke replied:
I expanded the docs for both `read` and `write` a bit.

alexcrichton commented on crates/wasmtime/src/externals.rs:
```diff
@@ -864,6 +902,38 @@ impl Memory {
+    ///
+    /// Returns an error if the copy failed.
+    pub fn read(&self, offset: usize, buffer: &mut [u8]) -> Result<(), MemoryError> {
+        if offset + buffer.len() > self.data_size() {
```
I think here this will want to use `checked_add` to avoid overflow.

theduke replied:
Ah yeah, I'm too used to `debug` semantics for overflow checking...

Both `read` and `write` now use `checked_add`.

alexcrichton reviewed (commented):
Thanks for this!

alexcrichton commented on crates/wasmtime/src/externals.rs:
```diff
@@ -598,6 +598,21 @@ impl Table {
 }
 
+/// Error for out of bounds [`Memory`] access.
+#[derive(Debug, PartialEq)]
```
I think it should be ok to drop the `PartialEq` here

alexcrichton commented on crates/wasmtime/src/externals.rs:
```diff
@@ -864,6 +894,51 @@ impl Memory {
+    /// If the offset + buffer length would exceed the current memory capacity,
+    /// a [`MemoryAccessError`] is returned.
+    pub fn read(&self, offset: usize, buffer: &mut [u8]) -> Result<(), MemoryAccessError> {
+        // Guard against overflows.
```
It's ok to remove this comment, I don't think it adds much over what the existence of the `checked_add` already implies.

alexcrichton replied:
Although thinking about this again, another perhaps safer way to write these methods would be:

```rust
let src = self.data_unchecked().get(offset..offset + buffer.len())?;
buffer.copy_from_slice(src);
Ok(())
```

(and something similar below)

alexcrichton replied:
Or actually to avoid having to do checked arithmetic:

```rust
let src = self.data_unchecked().get(offset..).get(..buffer.len())?;
```

alexcrichton reviewed (commented):
I think there may be some rustfmt issues too?

theduke commented:
I implemented your suggestion, the code definitely is much more elegant this way.

I also added a `_private` field to the error for extensibility and moved the test to `tests/all/externals.rs` since most tests seem to live there.

alexcrichton commented:
Looks great to me, thanks!
